core/src/test/scala/org/apache/spark/shuffle/ShuffleSuite.scala

```diff
@@ -315,8 +320,115 @@ abstract class ShuffleSuite extends SparkFunSuite with Matchers with LocalSparkC
     assert(metrics.bytesWritten === metrics.byresRead)
     assert(metrics.bytesWritten > 0)
   }
+
+  def multipleAttemptConfs: Seq[(String, SparkConf)] = Seq("basic" -> conf)
+
+  multipleAttemptConfs.foreach { case (name, multipleAttemptConf) =>
+    test("multiple attempts for one task: conf = " + name) {
+      sc = new SparkContext("local", "test", multipleAttemptConf)
```

Comment:
This is the most important new test.  It makes sure that we can have multiple shuffle writers executing concurrently, and the output of both can be read.

`multipleAttempConfs` is just to make sure we can test different code paths from different confs, in particular in the `Unsafe` shuffle.